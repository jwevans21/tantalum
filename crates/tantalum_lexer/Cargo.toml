[package]
name = "tantalum_lexer"
version = "0.1.0"
authors.workspace = true
edition.workspace = true
license.workspace = true
readme.workspace = true
repository.workspace = true
publish.workspace = true

[dependencies]
serde = { workspace = true, optional = true }

tantalum_span = { path = "../tantalum_span" }

[features]
default = ["serde"]
serde = ["dep:serde"]

[dev-dependencies]
insta = { workspace = true}
pretty_assertions = { workspace = true }
rstest = { workspace = true }


[lints]
workspace = true
