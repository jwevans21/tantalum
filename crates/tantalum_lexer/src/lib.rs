//! # Tantalum Lexer
//!
//! This crate provides a lexer for the Tantalum language.

use tantalum_source::{SourceFile, SourceLocation, SourceSpan};
use tantalum_syntax::{SyntaxKind, SyntaxToken};

/// A lexer for the Tantalum language.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Lexer<'file> {
    /// The source file that the lexer is currently lexing.
    file: &'file SourceFile,
    /// The current location in the source file.
    location: SourceLocation,
}

impl<'file> Lexer<'file> {
    /// Create a new lexer from a source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let lexer = Lexer::new(file);
    /// ```
    #[must_use]
    pub fn new(file: &'file SourceFile) -> Self {
        Self {
            file,
            location: file.location_start(),
        }
    }

    /// Get the source file that the lexer is currently lexing.
    #[must_use]
    pub fn source(&self) -> &'file SourceFile {
        self.file
    }

    /// Get the current location in the source file.
    #[must_use]
    pub fn location(&self) -> SourceLocation {
        self.location
    }

    /// Get the character at a location in the source file.
    fn at_location(&self, location: SourceLocation) -> Option<char> {
        self.file.read_char(location)
    }

    /// Extract the next token from the source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    /// use tantalum_syntax::SyntaxKind;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let mut lexer = Lexer::new(file);
    ///
    /// let token = lexer.next_token().expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Fn);
    /// ```
    #[must_use]
    pub fn next_token(&mut self) -> Option<SyntaxToken> {
        let start = self.location;

        let (kind, end) = self.characters_to_token(start)?;

        self.location = end;

        Some(SyntaxToken::new(
            kind,
            SourceSpan::new(self.file.id(), start.offset(), end.offset()),
        ))
    }

    /// Peek at the next token in the source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    /// use tantalum_syntax::SyntaxKind;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let mut lexer = Lexer::new(file);
    ///
    /// let token = lexer.peek_token(1).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Fn);
    ///
    /// let token = lexer.peek_token(2).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Whitespace);
    ///
    /// let token = lexer.peek_token(3).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Identifier);
    /// ```
    #[must_use]
    pub fn peek_token(&self, mut n: usize) -> Option<SyntaxToken> {
        let mut start = self.location;

        let (kind, end) = loop {
            n -= 1;
            let (kind, end) = self.characters_to_token(start)?;

            if n == 0 {
                break (kind, end);
            }

            start = end;
        };

        Some(SyntaxToken::new(
            kind,
            SourceSpan::new(self.file.id(), start.offset(), end.offset()),
        ))
    }

    /// Convert a character to a [`SyntaxKind`].
    ///
    /// [`SyntaxKind`]: tantalum_syntax::SyntaxKind
    fn characters_to_token(&self, start: SourceLocation) -> Option<(SyntaxKind, SourceLocation)> {
        let mut current = start;

        let c = self.at_location(current)?;
        current = current.next_by(c.len_utf8());

        let token = match c {
            '(' => SyntaxKind::LParen,
            ')' => SyntaxKind::RParen,
            '{' => SyntaxKind::LBrace,
            '}' => SyntaxKind::RBrace,
            '[' => SyntaxKind::LBracket,
            ']' => SyntaxKind::RBracket,
            '<' => SyntaxKind::LAngle,
            '>' => SyntaxKind::RAngle,
            ',' => SyntaxKind::Comma,
            ':' => SyntaxKind::Colon,
            ';' => SyntaxKind::Semicolon,
            '.' => SyntaxKind::Dot,
            '&' => SyntaxKind::Ampersand,
            '|' => SyntaxKind::Pipe,
            '^' => SyntaxKind::Caret,
            '+' => SyntaxKind::Plus,
            '-' => SyntaxKind::Minus,
            '*' => SyntaxKind::Star,
            '/' => SyntaxKind::Slash,
            '\\' => SyntaxKind::Backslash,
            '%' => SyntaxKind::Percent,
            '=' => SyntaxKind::Equals,
            '!' => SyntaxKind::Bang,
            '~' => SyntaxKind::Tilde,
            '?' => SyntaxKind::Question,
            '@' => SyntaxKind::At,
            '0'..='9' => SyntaxKind::Digit,
            '_' | 'a'..='z' | 'A'..='Z' => {
                while let Some(c) = self.at_location(current) {
                    match c {
                        '_' | 'a'..='z' | 'A'..='Z' | '0'..='9' => {
                            current = current.next_by(c.len_utf8());
                        }
                        _ => break,
                    }
                }

                match self.file.read_span(SourceSpan::new(
                    self.file.id(),
                    self.location.offset(),
                    current.offset(),
                ))? {
                    "fn" => SyntaxKind::Fn,
                    "let" => SyntaxKind::Let,
                    "return" => SyntaxKind::Return,
                    "if" => SyntaxKind::If,
                    "else" => SyntaxKind::Else,
                    "true" => SyntaxKind::True,
                    "false" => SyntaxKind::False,
                    _ => SyntaxKind::Identifier,
                }
            }
            c if c.is_whitespace() => SyntaxKind::Whitespace,
            _ => SyntaxKind::Other,
        };

        Some((token, current))
    }
}

#[test]
fn keywords() {
    use crate::Lexer;
    use tantalum_source::SourceFileCollection;
    use tantalum_syntax::SyntaxKind;

    let mut files = SourceFileCollection::new();

    let file_id = files.add_file(
        "file.tan".to_string(),
        "fn let return if else true false".to_string(),
    );

    let file = files
        .get_file(file_id)
        .expect("file should exist since it was just added");

    let mut lexer = Lexer::new(file);

    let mut tokens = Vec::new();
    while let Some(token) = lexer.next_token() {
        if token.kind() == SyntaxKind::Whitespace {
            continue;
        }
        tokens.push(token);
    }

    assert_eq!(tokens.len(), 7);

    assert_eq!(tokens[0].kind(), SyntaxKind::Fn);
    assert_eq!(tokens[1].kind(), SyntaxKind::Let);
    assert_eq!(tokens[2].kind(), SyntaxKind::Return);
    assert_eq!(tokens[3].kind(), SyntaxKind::If);
    assert_eq!(tokens[4].kind(), SyntaxKind::Else);
    assert_eq!(tokens[5].kind(), SyntaxKind::True);
    assert_eq!(tokens[6].kind(), SyntaxKind::False);
}

#[test]
fn identifiers() {
    use crate::Lexer;
    use tantalum_source::SourceFileCollection;
    use tantalum_syntax::SyntaxKind;

    let mut files = SourceFileCollection::new();

    let file_id = files.add_file(
        "file.tan".to_string(),
        "a A _ a1 A1 _1 _a _A __AA01".to_string(),
    );

    let file = files
        .get_file(file_id)
        .expect("file should exist since it was just added");

    let mut lexer = Lexer::new(file);

    let mut tokens = Vec::new();
    while let Some(token) = lexer.next_token() {
        if token.kind() == SyntaxKind::Whitespace {
            continue;
        }
        tokens.push(token);
    }

    assert_eq!(tokens.len(), 9);

    assert_eq!(tokens[0].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[1].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[2].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[3].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[4].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[5].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[6].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[7].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[8].kind(), SyntaxKind::Identifier);
}
