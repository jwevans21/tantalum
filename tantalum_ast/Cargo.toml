[package]
name = "tantalum_ast"
version = "0.1.0"
authors.workspace = true
edition.workspace = true
license.workspace = true
readme.workspace = true
repository.workspace = true
publish.workspace = true

[dependencies]
serde = { workspace = true, optional = true }

tantalum_span = { path = "../tantalum_span" }

[features]
default = ["serde"]
serde = ["dep:serde"]

[lints]
workspace = true

