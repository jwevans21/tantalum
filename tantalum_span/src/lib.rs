//! # Tantalum Span
//!
//! Provides a span to locate the positions and ranges of tokens in a file.

use core::{
    fmt::{Debug, Display, Formatter, Result as FmtResult},
    num::Saturating,
    ops::{Deref, Range},
};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Location<'file_name> {
    /// The name of the file that this location is in
    #[cfg_attr(feature = "serde", serde(borrow))]
    file_name: &'file_name str,
    /// The byte of the location in the file
    position: Saturating<usize>,
    /// The line number of the location in the file
    line: Saturating<usize>,
    /// The column number of the location in the file
    column: Saturating<usize>,
}

impl<'file_name> Location<'file_name> {
    #[must_use]
    #[inline]
    pub fn new(file_name: &'file_name str) -> Self {
        return Self {
            file_name,
            position: Saturating(0),
            line: Saturating(1),
            column: Saturating(1),
        };
    }

    #[must_use]
    #[inline]
    pub fn new_at(file_name: &'file_name str, position: usize, line: usize, column: usize) -> Self {
        return Self {
            file_name,
            position: Saturating(position),
            line: Saturating(line),
            column: Saturating(column),
        };
    }

    #[inline]
    pub fn advance(&mut self, character: char) {
        if character == '\n' {
            self.line += 1;
            self.column = Saturating(1);
        } else {
            self.column += 1;
        }

        self.position += character.len_utf8();
    }

    #[must_use]
    #[inline]
    pub fn file_name(&self) -> &'file_name str {
        return self.file_name;
    }

    #[must_use]
    #[inline]
    pub fn position(&self) -> usize {
        return self.position.0;
    }

    #[must_use]
    #[inline]
    pub fn line(&self) -> usize {
        return self.line.0;
    }

    #[must_use]
    #[inline]
    pub fn column(&self) -> usize {
        return self.column.0;
    }

    #[must_use]
    #[inline]
    pub fn range(&self, other: &Self) -> Range<usize> {
        return (self.position())..(other.position());
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Span<'file_name> {
    /// The starting location of the span
    #[cfg_attr(feature = "serde", serde(borrow))]
    start: Location<'file_name>,
    /// The ending location of the span
    end: Location<'file_name>,
}

impl<'file_name> Span<'file_name> {
    #[must_use]
    #[inline]
    pub fn new(start: Location<'file_name>, end: Location<'file_name>) -> Self {
        return Self { start, end };
    }

    #[must_use]
    #[inline]
    pub fn file_name(&self) -> &'file_name str {
        return self.start.file_name();
    }

    #[must_use]
    #[inline]
    pub fn range(&self) -> Range<usize> {
        return self.start.range(&self.end);
    }

    #[must_use]
    #[inline]
    pub fn start(&self) -> Location<'file_name> {
        return self.start;
    }

    #[must_use]
    #[inline]
    pub fn end(&self) -> Location<'file_name> {
        return self.end;
    }

    #[must_use]
    #[inline]
    pub fn line(&self) -> usize {
        return self.start.line();
    }

    #[must_use]
    #[inline]
    pub fn column(&self) -> usize {
        return self.start.column();
    }
}

impl Display for Span<'_> {
    #[inline]
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        return write!(f, "{}..{}", self.start.position(), self.end.position());
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Spanned<'file_name, T: Debug + Clone + PartialEq + Eq> {
    #[cfg_attr(feature = "serde", serde(borrow))]
    span: Span<'file_name>,
    data: T,
}

impl<'file_name, T> Spanned<'file_name, T>
where
    T: Debug + Clone + PartialEq + Eq,
{
    #[must_use]
    #[inline]
    pub fn new(span: Span<'file_name>, data: T) -> Self {
        return Self { span, data };
    }

    #[must_use]
    #[inline]
    pub fn spanning(start: Location<'file_name>, end: Location<'file_name>, data: T) -> Self {
        return Self::new(Span::new(start, end), data);
    }

    #[must_use]
    #[inline]
    pub fn join_spans(start: Span<'file_name>, end: Span<'file_name>, data: T) -> Self {
        return Self::new(Span::new(start.start, end.end), data);
    }

    pub fn map<U, F>(self, f: F) -> Spanned<'file_name, U>
    where
        U: Debug + Clone + PartialEq + Eq,
        F: FnOnce(T) -> U,
    {
        return Spanned::new(self.span, f(self.data));
    }

    #[must_use]
    #[inline]
    pub fn start(&self) -> Location<'file_name> {
        return self.span.start();
    }

    #[must_use]
    #[inline]
    pub fn end(&self) -> Location<'file_name> {
        return self.span.end();
    }

    #[must_use]
    #[inline]
    pub fn data(&self) -> &T {
        return &self.data;
    }

    #[must_use]
    #[inline]
    pub fn span(&self) -> Span<'file_name> {
        return self.span;
    }

    #[must_use]
    #[inline]
    pub fn range(&self) -> Range<usize> {
        return self.span.range();
    }

    #[must_use]
    #[inline]
    pub fn line(&self) -> usize {
        return self.span.line();
    }

    #[must_use]
    #[inline]
    pub fn column(&self) -> usize {
        return self.span.column();
    }

    #[must_use]
    #[inline]
    pub fn file_name(&self) -> &'file_name str {
        return self.span.file_name();
    }
}

impl<T> Copy for Spanned<'_, T> where T: Copy + Debug + Clone + PartialEq + Eq {}

impl<T> AsRef<T> for Spanned<'_, T>
where
    T: Debug + Clone + PartialEq + Eq,
{
    #[inline]
    fn as_ref(&self) -> &T {
        return &self.data;
    }
}

impl<T> Deref for Spanned<'_, T>
where
    T: Debug + Clone + PartialEq + Eq,
{
    type Target = T;

    #[inline]
    fn deref(&self) -> &Self::Target {
        return &self.data;
    }
}

impl<T> Display for Spanned<'_, T>
where
    T: Display + Debug + Clone + PartialEq + Eq,
{
    #[inline]
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        return write!(f, "{} @ {}", self.data, self.span);
    }
}
