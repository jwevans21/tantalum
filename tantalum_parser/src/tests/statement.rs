use tantalum_ast::{
    BinaryOperator, BinaryOperatorKind, Expression, ExpressionKind, Statement, StatementKind, Type,
    TypeKind,
};
use tantalum_lexer::Lexer;
use tantalum_span::Span;

use pretty_assertions::assert_eq;

use crate::Parser;

#[test]
fn expression_statement() {
    let source = "42 + 42;";
    let mut parser = Parser::new(Lexer::new("expression_statement", source));

    let result = parser.parse_statement();

    assert_eq!(
        result,
        Ok(Statement {
            span: Span::new("expression_statement", 0, 8, 1, 1),
            kind: StatementKind::Expression {
                expression: Expression {
                    span: Span::new("expression_statement", 0, 7, 1, 1),
                    kind: ExpressionKind::BinaryOperation {
                        left: Box::new(Expression {
                            span: Span::new("expression_statement", 0, 2, 1, 1),
                            kind: ExpressionKind::IntegerLiteral {
                                value: "42",
                                radix: 10
                            },
                        }),
                        operator: BinaryOperator {
                            span: Span::new("expression_statement", 3, 4, 1, 4),
                            kind: BinaryOperatorKind::Addition,
                        },
                        right: Box::new(Expression {
                            span: Span::new("expression_statement", 5, 7, 1, 6),
                            kind: ExpressionKind::IntegerLiteral {
                                value: "42",
                                radix: 10
                            },
                        }),
                    },
                },
            },
        })
    );
}

#[test]
fn let_statement() {
    let source = "let x = 42;";
    let mut parser = Parser::new(Lexer::new("let_statement", source));

    let result = parser.parse_statement();

    assert_eq!(
        result,
        Ok(Statement {
            span: Span::new("let_statement", 0, 11, 1, 1),
            kind: StatementKind::VariableDeclaration {
                name: "x",
                ty: None,
                value: Expression {
                    span: Span::new("let_statement", 8, 10, 1, 9),
                    kind: ExpressionKind::IntegerLiteral {
                        value: "42",
                        radix: 10
                    },
                },
            },
        })
    );
}

#[test]
fn let_statement_with_type() {
    let source = "let x: *u8 = \"Hello, World!\";";
    let mut parser = Parser::new(Lexer::new("let_statement", source));

    let result = parser.parse_statement();

    assert_eq!(
        result,
        Ok(Statement {
            span: Span::new("let_statement", 0, 29, 1, 1),
            kind: StatementKind::VariableDeclaration {
                name: "x",
                ty: Some(Type {
                    span: Span::new("let_statement", 7, 10, 1, 8),
                    kind: TypeKind::Pointer(Box::new(Type {
                        span: Span::new("let_statement", 8, 10, 1, 9),
                        kind: TypeKind::Named("u8"),
                    })),
                }),
                value: Expression {
                    span: Span::new("let_statement", 13, 28, 1, 14),
                    kind: ExpressionKind::StringLiteral {
                        value: "\"Hello, World!\""
                    },
                },
            },
        })
    );
}

#[test]
fn if_statement() {
    let source = "if true { return 42; }";
    let mut parser = Parser::new(Lexer::new("if_statement", source));

    let result = parser.parse_statement();

    assert_eq!(
        result,
        Ok(Statement {
            span: Span::new("if_statement", 0, 22, 1, 1),
            kind: StatementKind::If {
                condition: Expression {
                    span: Span::new("if_statement", 3, 7, 1, 4),
                    kind: ExpressionKind::BooleanLiteral { value: true },
                },
                body: Box::new(Statement {
                    span: Span::new("if_statement", 8, 22, 1, 9),
                    kind: StatementKind::Block {
                        statements: vec![Statement {
                            span: Span::new("if_statement", 10, 20, 1, 11),
                            kind: StatementKind::Return {
                                value: Some(Expression {
                                    span: Span::new("if_statement", 17, 19, 1, 18),
                                    kind: ExpressionKind::IntegerLiteral {
                                        value: "42",
                                        radix: 10
                                    },
                                }),
                            },
                        }],
                    },
                }),
                else_branch: None,
            },
        })
    );
}

#[test]
fn if_else_statement() {
    let source = "if true { return 42; } else { return 0; }";
    let mut parser = Parser::new(Lexer::new("if_else_statement", source));

    let result = parser.parse_statement();

    assert_eq!(
        result,
        Ok(Statement {
            span: Span::new("if_else_statement", 0, 41, 1, 1),
            kind: StatementKind::If {
                condition: Expression {
                    span: Span::new("if_else_statement", 3, 7, 1, 4),
                    kind: ExpressionKind::BooleanLiteral { value: true },
                },
                body: Box::new(Statement {
                    span: Span::new("if_else_statement", 8, 22, 1, 9),
                    kind: StatementKind::Block {
                        statements: vec![Statement {
                            span: Span::new("if_else_statement", 10, 20, 1, 11),
                            kind: StatementKind::Return {
                                value: Some(Expression {
                                    span: Span::new("if_else_statement", 17, 19, 1, 18),
                                    kind: ExpressionKind::IntegerLiteral {
                                        value: "42",
                                        radix: 10
                                    },
                                }),
                            },
                        }],
                    },
                }),
                else_branch: Some(Box::new(Statement {
                    span: Span::new("if_else_statement", 28, 41, 1, 29),
                    kind: StatementKind::Block {
                        statements: vec![Statement {
                            span: Span::new("if_else_statement", 30, 39, 1, 31),
                            kind: StatementKind::Return {
                                value: Some(Expression {
                                    span: Span::new("if_else_statement", 37, 38, 1, 38),
                                    kind: ExpressionKind::IntegerLiteral {
                                        value: "0",
                                        radix: 10
                                    },
                                }),
                            },
                        }],
                    },
                })),
            },
        })
    );
}

#[test]
#[allow(clippy::too_many_lines)]
fn for_loop_statement() {
    let source = "for let i = 0; i < 10; i = i + 1; { return i; }";
    let mut parser = Parser::new(Lexer::new("for_loop_statement", source));

    let result = parser.parse_statement();

    assert_eq!(
        result,
        Ok(Statement {
            span: Span::new("for_loop_statement", 0, 47, 1, 1),
            kind: StatementKind::ForInitCondUpdate {
                initialization: Box::new(Statement {
                    span: Span::new("for_loop_statement", 4, 14, 1, 5),
                    kind: StatementKind::VariableDeclaration {
                        name: "i",
                        ty: None,
                        value: Expression {
                            span: Span::new("for_loop_statement", 12, 13, 1, 13),
                            kind: ExpressionKind::IntegerLiteral {
                                value: "0",
                                radix: 10,
                            },
                        },
                    },
                }),
                condition: Box::new(Statement {
                    span: Span::new("for_loop_statement", 15, 22, 1, 16),
                    kind: StatementKind::Expression {
                        expression: Expression {
                            span: Span::new("for_loop_statement", 15, 21, 1, 16),
                            kind: ExpressionKind::BinaryOperation {
                                left: Box::new(Expression {
                                    span: Span::new("for_loop_statement", 15, 16, 1, 16),
                                    kind: ExpressionKind::Variable { name: "i" },
                                }),
                                operator: BinaryOperator {
                                    span: Span::new("for_loop_statement", 17, 18, 1, 18),
                                    kind: BinaryOperatorKind::LessThan,
                                },
                                right: Box::new(Expression {
                                    span: Span::new("for_loop_statement", 19, 21, 1, 20),
                                    kind: ExpressionKind::IntegerLiteral {
                                        value: "10",
                                        radix: 10,
                                    },
                                }),
                            },
                        },
                    },
                }),
                update: Box::new(Statement {
                    span: Span::new("for_loop_statement", 23, 33, 1, 24),
                    kind: StatementKind::Expression {
                        expression: Expression {
                            span: Span::new("for_loop_statement", 23, 32, 1, 24),
                            kind: ExpressionKind::BinaryOperation {
                                left: Box::new(Expression {
                                    span: Span::new("for_loop_statement", 23, 24, 1, 24),
                                    kind: ExpressionKind::Variable { name: "i" },
                                }),
                                operator: BinaryOperator {
                                    span: Span::new("for_loop_statement", 25, 26, 1, 26),
                                    kind: BinaryOperatorKind::Assignment,
                                },
                                right: Box::new(Expression {
                                    span: Span::new("for_loop_statement", 27, 32, 1, 28),
                                    kind: ExpressionKind::BinaryOperation {
                                        left: Box::new(Expression {
                                            span: Span::new("for_loop_statement", 27, 28, 1, 28),
                                            kind: ExpressionKind::Variable { name: "i" },
                                        }),
                                        operator: BinaryOperator {
                                            span: Span::new("for_loop_statement", 29, 30, 1, 30),
                                            kind: BinaryOperatorKind::Addition,
                                        },
                                        right: Box::new(Expression {
                                            span: Span::new("for_loop_statement", 31, 32, 1, 32),
                                            kind: ExpressionKind::IntegerLiteral {
                                                value: "1",
                                                radix: 10,
                                            },
                                        }),
                                    },
                                }),
                            },
                        },
                    },
                }),
                body: Box::new(Statement {
                    span: Span::new("for_loop_statement", 34, 47, 1, 35),
                    kind: StatementKind::Block {
                        statements: vec![Statement {
                            span: Span::new("for_loop_statement", 36, 45, 1, 37),
                            kind: StatementKind::Return {
                                value: Some(Expression {
                                    span: Span::new("for_loop_statement", 43, 44, 1, 44),
                                    kind: ExpressionKind::Variable { name: "i" },
                                }),
                            },
                        }],
                    },
                }),
            }
        })
    );
}

#[test]
fn while_statement() {
    let source = "while i < 10 { i = i + 1; }";
    let mut parser = Parser::new(Lexer::new("while_statement", source));

    let result = parser.parse_statement();

    assert_eq!(
        result,
        Ok(Statement {
            span: Span::new("while_statement", 0, 27, 1, 1),
            kind: StatementKind::While {
                condition: Expression {
                    span: Span::new("while_statement", 6, 12, 1, 7),
                    kind: ExpressionKind::BinaryOperation {
                        left: Box::new(Expression {
                            span: Span::new("while_statement", 6, 7, 1, 7),
                            kind: ExpressionKind::Variable { name: "i" },
                        }),
                        operator: BinaryOperator {
                            span: Span::new("while_statement", 8, 9, 1, 9),
                            kind: BinaryOperatorKind::LessThan,
                        },
                        right: Box::new(Expression {
                            span: Span::new("while_statement", 10, 12, 1, 11),
                            kind: ExpressionKind::IntegerLiteral {
                                value: "10",
                                radix: 10,
                            },
                        }),
                    },
                },
                body: Box::new(Statement {
                    span: Span::new("while_statement", 13, 27, 1, 14),
                    kind: StatementKind::Block {
                        statements: vec![Statement {
                            span: Span::new("while_statement", 15, 25, 1, 16),
                            kind: StatementKind::Expression {
                                expression: Expression {
                                    span: Span::new("while_statement", 15, 24, 1, 16),
                                    kind: ExpressionKind::BinaryOperation {
                                        left: Box::new(Expression {
                                            span: Span::new("while_statement", 15, 16, 1, 16),
                                            kind: ExpressionKind::Variable { name: "i" },
                                        }),
                                        operator: BinaryOperator {
                                            span: Span::new("while_statement", 17, 18, 1, 18),
                                            kind: BinaryOperatorKind::Assignment,
                                        },
                                        right: Box::new(Expression {
                                            span: Span::new("while_statement", 19, 24, 1, 20),
                                            kind: ExpressionKind::BinaryOperation {
                                                left: Box::new(Expression {
                                                    span: Span::new(
                                                        "while_statement",
                                                        19,
                                                        20,
                                                        1,
                                                        20
                                                    ),
                                                    kind: ExpressionKind::Variable { name: "i" },
                                                }),
                                                operator: BinaryOperator {
                                                    span: Span::new(
                                                        "while_statement",
                                                        21,
                                                        22,
                                                        1,
                                                        22
                                                    ),
                                                    kind: BinaryOperatorKind::Addition,
                                                },
                                                right: Box::new(Expression {
                                                    span: Span::new(
                                                        "while_statement",
                                                        23,
                                                        24,
                                                        1,
                                                        24
                                                    ),
                                                    kind: ExpressionKind::IntegerLiteral {
                                                        value: "1",
                                                        radix: 10,
                                                    },
                                                }),
                                            },
                                        }),
                                    },
                                },
                            },
                        }],
                    },
                }),
            },
        })
    );
}

#[test]
fn return_statement_integer() {
    let source = "return 42;";
    let mut parser = Parser::new(Lexer::new("return_statement", source));

    let result = parser.parse_statement();

    assert_eq!(
        result,
        Ok(Statement {
            span: Span::new("return_statement", 0, 10, 1, 1),
            kind: StatementKind::Return {
                value: Some(Expression {
                    span: Span::new("return_statement", 7, 9, 1, 8),
                    kind: ExpressionKind::IntegerLiteral {
                        value: "42",
                        radix: 10
                    },
                },),
            },
        })
    );
}

#[test]
fn return_statement_expression() {
    let source = "return 42 + 42;";
    let mut parser = Parser::new(Lexer::new("return_statement", source));

    let result = parser.parse_statement();

    assert_eq!(
        result,
        Ok(Statement {
            span: Span::new("return_statement", 0, 15, 1, 1),
            kind: StatementKind::Return {
                value: Some(Expression {
                    span: Span::new("return_statement", 7, 14, 1, 8),
                    kind: ExpressionKind::BinaryOperation {
                        left: Box::new(Expression {
                            span: Span::new("return_statement", 7, 9, 1, 8),
                            kind: ExpressionKind::IntegerLiteral {
                                value: "42",
                                radix: 10
                            },
                        }),
                        operator: BinaryOperator {
                            span: Span::new("return_statement", 10, 11, 1, 11),
                            kind: BinaryOperatorKind::Addition,
                        },
                        right: Box::new(Expression {
                            span: Span::new("return_statement", 12, 14, 1, 13),
                            kind: ExpressionKind::IntegerLiteral {
                                value: "42",
                                radix: 10
                            },
                        }),
                    },
                },),
            },
        })
    );
}

#[test]
fn return_statement_no_expression() {
    let source = "return;";
    let mut parser = Parser::new(Lexer::new("return_statement", source));

    let result = parser.parse_statement();

    assert_eq!(
        result,
        Ok(Statement {
            span: Span::new("return_statement", 0, 7, 1, 1),
            kind: StatementKind::Return { value: None },
        })
    );
}

#[test]
fn break_statement() {
    let source = "break;";
    let mut parser = Parser::new(Lexer::new("break_statement", source));

    let result = parser.parse_statement();

    assert_eq!(
        result,
        Ok(Statement {
            span: Span::new("break_statement", 0, 6, 1, 1),
            kind: StatementKind::Break,
        })
    );
}

#[test]
fn continue_statement() {
    let source = "continue;";
    let mut parser = Parser::new(Lexer::new("continue_statement", source));

    let result = parser.parse_statement();

    assert_eq!(
        result,
        Ok(Statement {
            span: Span::new("continue_statement", 0, 9, 1, 1),
            kind: StatementKind::Continue,
        })
    );
}

#[test]
fn block_statement() {
    let source = "{ let x = 42; }";
    let mut parser = Parser::new(Lexer::new("block_statement", source));

    let result = parser.parse_statement();

    assert_eq!(
        result,
        Ok(Statement {
            span: Span::new("block_statement", 0, 15, 1, 1),
            kind: StatementKind::Block {
                statements: vec![Statement {
                    span: Span::new("block_statement", 2, 13, 1, 3),
                    kind: StatementKind::VariableDeclaration {
                        name: "x",
                        ty: None,
                        value: Expression {
                            span: Span::new("block_statement", 10, 12, 1, 11),
                            kind: ExpressionKind::IntegerLiteral {
                                value: "42",
                                radix: 10
                            },
                        },
                    },
                }],
            },
        })
    );
}
