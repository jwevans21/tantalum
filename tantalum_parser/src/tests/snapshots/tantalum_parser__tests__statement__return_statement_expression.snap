---
source: tantalum_parser/src/tests/statement.rs
expression: result
snapshot_kind: text
---
Ok(Statement(
  span: Span(
    start: Location(
      file_name: "return_statement",
      position: 0,
      line: 1,
      column: 1,
    ),
    end: Location(
      file_name: "return_statement",
      position: 15,
      line: 1,
      column: 16,
    ),
  ),
  kind: Return(
    value: Some(Expression(
      span: Span(
        start: Location(
          file_name: "return_statement",
          position: 7,
          line: 1,
          column: 8,
        ),
        end: Location(
          file_name: "return_statement",
          position: 14,
          line: 1,
          column: 15,
        ),
      ),
      kind: BinaryOperation(
        left: Expression(
          span: Span(
            start: Location(
              file_name: "return_statement",
              position: 7,
              line: 1,
              column: 8,
            ),
            end: Location(
              file_name: "return_statement",
              position: 9,
              line: 1,
              column: 10,
            ),
          ),
          kind: IntegerLiteral(
            value: "42",
            radix: 10,
          ),
        ),
        operator: BinaryOperator(
          span: Span(
            start: Location(
              file_name: "return_statement",
              position: 10,
              line: 1,
              column: 11,
            ),
            end: Location(
              file_name: "return_statement",
              position: 11,
              line: 1,
              column: 12,
            ),
          ),
          kind: Addition,
        ),
        right: Expression(
          span: Span(
            start: Location(
              file_name: "return_statement",
              position: 12,
              line: 1,
              column: 13,
            ),
            end: Location(
              file_name: "return_statement",
              position: 14,
              line: 1,
              column: 15,
            ),
          ),
          kind: IntegerLiteral(
            value: "42",
            radix: 10,
          ),
        ),
      ),
    )),
  ),
))
